configs,stdlib,tests: Remove get_runtime_isa()

BobbyRBruce opened the pull request:
`get_runtime_isa()` has been deprecated for some time. It is a leftover piece of code from when gem5 was compiled to a single ISA and that ISA used to configure the simulated system to use that ISA. Since multi-ISA compilations are possible, `get_runtime_isa()` should not be used. Unless the gem5 binary is compiled to a single ISA, a failure will occur.

The new proceedure for specify which ISA to use is by the setting of the correct `BaseCPU` implementation. E.g., `X86SimpleTimingCPU` of `ArmO3CPU`.

This patch removes the remaining `get_runtime_isa()` instances and removes the function itself. The `SimpleCore` class has been updated to allow for it's CPU factory to return a class, needed by scripts in "configs/common".

The deprecated functionality in the standard library, which allowed for the specifying of an ISA when setting up a processor and/or core has also been removed. Setting an ISA is now manditory.

Fixes #216.

BobbyRBruce committed:
configs,stdlib,tests: Remove get_runtime_isa()

`get_runtime_isa()` has been deprecated for some time. It is a leftover
piece of code from when gem5 was compiled to a single ISA and that ISA
used to configure the simulated system to use that ISA. Since multi-ISA
compilations are possible, `get_runtime_isa()` should not be used.
Unless the gem5 binary is compiled to a single ISA, a failure will
occur.

The new proceedure for specify which ISA to use is by the setting of the
correct `BaseCPU` implementation. E.g., `X86SimpleTimingCPU` of
`ArmO3CPU`.

This patch removes the remaining `get_runtime_isa()` instances and
removes the function itself. The `SimpleCore` class has been updated to
allow for it's CPU factory to return a class, needed by scripts in
"configs/common".

The deprecated functionality in the standard library, which allowed for
the specifying of an ISA when setting up a processor and/or core has
also been removed. Setting an ISA is now manditory.

Issue-on: https://github.com/gem5/gem5/issues/216 : Shows the se.py
cannot work with multi-ISA gem5 builds.

Change-Id: Iebfc6c69a7d40e61228e03e1423c1c3c7f407f80

BobbyRBruce commented:
Just a draft for now. Think this solution is almost there but haven't tested it at all.

giactra commented:
Hi Bobby, thanks for providing a PR. Let me know when you want me to review it (I see it still marked as a draft)

giactra commented on configs/common/Options.py:
```diff
@@ -254,6 +256,13 @@ def addCommonOptions(parser):
         help="type of cpu to run with",
     )
+
+    parser.add_argument(
+        "--isa",
+        choices=get_isas_str_set(),
+        help="The ISA the simulated system will use",
+    )
```
How would it work with a multi-ISA simulation? It seems you can only specify one ISA with this option

BobbyRBruce replied:
The option is just a single setting yes, but we could expand it anyway you want to permit more. Ultimately as long as you're compiled with the ISA, we can use it. It's just selecting the right BaseCPU implementation.

giactra commented on configs/common/CacheConfig.py:
```diff
@@ -117,7 +116,7 @@ def config_cache(options, system):
 
-        if get_runtime_isa() in [ISA.X86, ISA.RISCV]:
+        if get_isa_from_str(options.isa) in [ISA.X86, ISA.RISCV]:
             walk_cache_class = PageTableWalkerCache
```
Can't we just remove the walk_cache_class variable and the if clause?
We can handle this down below during cache instantiation:

            if walk_cache_class:
                iwalkcache = walk_cache_class()
                dwalkcache = walk_cache_class()
            else:
                iwalkcache = None
                dwalkcache = None

We can just transform the snippe above and instantiate a PageTableWalker depending on the CPU isa.
Something like (might not be accurate):

```
if isinstance(system.cpus[i], (RISCVCpu, X86Cpu)):
    iwalkcache = PageTableWalkerCache(is_read_only=False,...)
    dwalkcache = PageTableWalkerCache(is_read_only=False,...):
else:
    iwalkcache = None
    dwalkcache = None
```

BobbyRBruce replied:
Totally agree with moving it, and have done so.

I forgot we're iterating on the CPU's here, so yes, `system.cpu[i]` is the safest way.

giactra commented on configs/common/CacheConfig.py:
```diff
@@ -192,7 +191,7 @@ def config_cache(options, system):
             # it to use this naming scheme; if this isn't possible, change
             # the names below.
-            if get_runtime_isa() in [ISA.X86, ISA.ARM, ISA.RISCV]:
+            if get_isa_from_str(options.isa) in [ISA.X86, ISA.ARM, ISA.RISCV]:
```
Also here I would use the system.cpu[i] type

BobbyRBruce replied:
Done

giactra reviewed (changes requested):
Thanks Bobby... it makes sense we provide an --isa option in order to generate the right CPU but IMHO once the CPU/ISA is generated we should rely on the SimObject itself when differentiating between different CPUs. In this way we make our code less dependant on the Options.py module which we want to deprecate anyway in the long term

BobbyRBruce commented on configs/common/Simulation.py:
```diff
@@ -52,9 +56,11 @@
+def getCPUClass(cpu_type: str, isa: str):
     """Returns the required cpu class and the mode of operation."""
-    cls = ObjectList.cpu_list.get(cpu_type)
+    cls = SimpleCore.cpu_class_factory(
+        cpu_type=get_cpu_type_from_str(cpu_type), isa=get_isa_from_str(isa)
+    )
```
Note: I'm aware that this isn't all that elegant, I should really move this function out of `SimpleCore` in the stdlib, but it works for now and didn't want to take on that right now.

It's probably worth providing some more factories to allow people to select and configure cores like this.

giactra replied:
Hi Bobby, it doesn't look unelegant to me, the only issue is that the ObjectList contains a wider set of cpus (from the configs subdirectory) like HPI as an example...

Couldn't we still rely on the ObjectList module and let the user choose the ISA implicitly via the CPU type? I haven't checked but I presume that when you build with ALL you will have something like

ArmAtomicCPU, ArmTmingCPU, [...], RiscvAtomicCPU, RiscvTimingCPU, [...]

Let me know if I am missing something

BobbyRBruce replied:
I've added code here to utilize ObjectList (which is kind of doing the same thing. My only worry here is we're hard-coding our naming scheme here. Like our CPU implementations having the name `{{ISA in Camel Case}}{{CPU Type}}`. If you're ok with this though, i'm fine with it too.

powerjg replied:
IMO, this is OK for the "legacy" configuration scripts. I *really* want to see us focus our effort solely on the standard library because it's going to be much easier to maintain going forward. So, I think it's OK to use relatively brittle things like assuming a particular naming scheme in these legacy scripts to let them limp along a little longer.

BobbyRBruce commented:
> Thanks Bobby... it makes sense we provide an --isa option in order to generate the right CPU but IMHO once the CPU/ISA is generated we should rely on the SimObject itself when differentiating between different CPUs. In this way we make our code less dependant on the Options.py module which we want to deprecate anyway in the long term

I agree. We shouldn't be passing around user parameters as much as we do. Once they setup what they were supposed to setup (The CPU, for example) then they shouldn't really be needed anymore.

I've made the changes where you suggested.

My (minor) complaint is the BaseCPU classes have no direct, clean way for them to tell you what ISA their using. When used to infer it via the gem5 (lone-ISA) build, which was fine, but now it's a bit annoying. Ideally for something as noteworthy as "What ISA does this core use?", we'd have a member function called "get_isa()". However, for now, we have figure it out through the class's inheritance. It's not especially hard but it bothers how hacky it is, and I doubt it's all that obvious to someone who doesn't fully grasp our Object-Oriented design. 

I made a comment about coming up with some functions to get more information out of the BaseCPU object. Something like `get_isa(cpu: BaseCPU) -> ISA` would be a good start if adding a "getter" to the class wasn't feasible.  Anyway, not really asking for anything. Just writing down my thought on this.

BobbyRBruce commented:
That's for the review @giactra . I wan't completely read when I took it out of draft status but I think it's fine now. I _think_ the tests will pass now... we'll see...

giactra commented on configs/common/Caches.py:
```diff
@@ -96,11 +95,11 @@ class PageTableWalkerCache(Cache):
+    # Used for cases where the target ISA isn't X86 or RISCV
+    is_read_only = True
+    # Writeback clean lines as well
+    writeback_clean = True
```
This one is not needed.. The other ISA which was using a PTW cache was Arm.. However I implemented a more realistic version some time ago... So this kind of cache is currently not used by anybody

giactra commented on configs/common/Simulation.py:
```diff
@@ -44,6 +44,10 @@
 
+from gem5.components.processors.simple_core import SimpleCore
+from gem5.isas import get_isa_from_str
+from gem5.prcesors.cpu_types import get_cpu_type_from_str
```
Is this a typo?

BobbyRBruce replied:
Yes

BobbyRBruce commented:
The threading here seems to have gotten very confused.

@giactra : I think I've addressed all your comments.

giactra commented:
> The threading here seems to have gotten very confused.
> 
> @giactra : I think I've addressed all your comments.

Thanks Bobby for addressing my comments. There is one though I believe it hasn't been addressed in full, which is:

>> Couldn't we still rely on the ObjectList module and let the user choose the ISA implicitly via the CPU type? I haven't checked but I presume that when you build with ALL you will have something like ArmAtomicCPU, ArmTmingCPU, [...], RiscvAtomicCPU, RiscvTimingCPU, [...]

While the PR is now using ObjectList, we are still relying on the --isa option. The problem with the code as of now is that it still does not support the entire list of CPU types, like HPI, O3_ARM_v7a_3 that do no start with the **capitalize(isa)** prefix.

My suggestion would be to remove the --isa option and to rely on the --cpu-type option only... 

For example if I compile with a single ISA (like Arm), I could choose one of the following

`(choose from 'ArmAtomicSimpleCPU', 'ArmMinorCPU', 'ArmNonCachingSimpleCPU', 'ArmO3CPU', 'ArmTimingSimpleCPU', 'AtomicSimpleCPU', 'BaseAtomicSimpleCPU', 'BaseMinorCPU', 'BaseNonCachingSimpleCPU', 'BaseO3CPU', 'BaseTimingSimpleCPU', 'DerivO3CPU', 'MinorCPU', 'NonCachingSimpleCPU', 'O3CPU', 'TimingSimpleCPU', 'HPI', 'O3_ARM_v7a_3', 'O3_ARM_v7a_3_Etrace', 'ex5_LITTLE', 'ex5_big')`

If I compile with multi-ISA (ALL), I could choose:

`(choose from 'ArmAtomicSimpleCPU', 'ArmMinorCPU', 'ArmNonCachingSimpleCPU', 'ArmO3CPU', 'ArmTimingSimpleCPU', 'BaseAtomicSimpleCPU', 'BaseMinorCPU', 'BaseNonCachingSimpleCPU', 'BaseO3CPU', 'BaseTimingSimpleCPU', 'MipsAtomicSimpleCPU', 'MipsMinorCPU', 'MipsNonCachingSimpleCPU', 'MipsO3CPU', 'MipsTimingSimpleCPU', 'PowerAtomicSimpleCPU', 'PowerMinorCPU', 'PowerNonCachingSimpleCPU', 'PowerO3CPU', 'PowerTimingSimpleCPU', 'RiscvAtomicSimpleCPU', 'RiscvMinorCPU', 'RiscvNonCachingSimpleCPU', 'RiscvO3CPU', 'RiscvTimingSimpleCPU', 'SparcAtomicSimpleCPU', 'SparcMinorCPU', 'SparcNonCachingSimpleCPU', 'SparcO3CPU', 'SparcTimingSimpleCPU', 'X86AtomicSimpleCPU', 'X86KvmCPU', 'X86MinorCPU', 'X86NonCachingSimpleCPU', 'X86O3CPU', 'X86TimingSimpleCPU', 'HPI', 'O3_ARM_v7a_3', 'O3_ARM_v7a_3_Etrace', 'ex5_LITTLE', 'ex5_big')`

In other words the ISA choice will be tightly coupled with the CPU I use. If I want to run a multi-ISA simulation with an X86 ISA and an out of order CPU I would do:

`build/ALL/gem5.opt configs/deprecated/example/fs.py --cpu-type X86O3CPU`

This IMHO has the advantage of being more concise, it is simple and does not require any prefixing trick :)
As a default option for --cpu-type, it is AtomicSimpleCPU now, you could just prefix it with the isa as string.
```

253     parser.add_argument(
254         "--cpu-type",                                                                                                                                     
255         default=get_isas_str_set() + "AtomicSimpleCPU",
256         choices=ObjectList.cpu_list.get_names(),
257         help="type of cpu to run with",
258     ) 

```

Please let me know what you think of it.

P.S Anyway OT we should probably filter out BaseCPU types from the ObjectList output

mattsinc reviewed (commented):
This seems fine from a GPU perspective, but just making sure you checked that at least the daily tests pass with it?

BobbyRBruce commented on configs/common/Simulation.py:
```diff
@@ -78,9 +78,10 @@ def setCPUClass(options):
-        TmpClass = AtomicSimpleCPU
-        test_mem_mode = "atomic"
+        pass
+        # CPUClass = TmpClass
+        # TmpClass = AtomicSimpleCPU
+        # test_mem_mode = "atomic"
```
Here we need to setup an AtomicSimpleCPU for whatever the "target isa" is.

BobbyRBruce commented on configs/deprecated/example/se.py:
```diff
@@ -142,27 +140,27 @@ def get_processes(args):
-                file=sys.stderr,
-            )
-            sys.exit(1)
+    # for app in apps:
+    #    try:
+    #        if get_isa_from_str(args.isa) == ISA.ARM:
+    #            exec(
+    #                "workload = %s('arm_%s', 'linux', '%s')"
+    #                % (app, args.arm_iset, args.spec_input)
+    #            )
+    #        else:
+    #            # TARGET_ISA has been removed, but this is missing a ], so it
+    #            # has incorrect syntax and wasn't being used anyway.
+    #            exec(
+    #                "workload = %s(buildEnv['TARGET_ISA', 'linux', '%s')"
+    #                % (app, args.spec_input)
+    #            )
+    #        multiprocesses.append(workload.makeProcess())
+    #    except:
+    #        print(
+    ##            f"Unable to find workload for {args.isa}: {app}",
+    #           file=sys.stderr,
+    #        )
+    #        sys.exit(1)
```
I really don't know what this code does but it behaves differently depending on whether your targeting the arm ISA or nto

BobbyRBruce commented on configs/example/hmc_hello.py:
```diff
@@ -68,10 +67,13 @@
-binary = "tests/test-progs/hello/bin/" + isa.name.lower() + "/linux/hello"
+
+
+# binary = (
+#    "tests/test-progs/hello/bin/" + options.isa.name.lower() + "/linux/hello"
+# )
```
Here is a problem if we remove the target ISA.

We could change this so we pass the binary as a parameter in this case. Is that ok?

BobbyRBruce commented on configs/ruby/Ruby.py:
```diff
@@ -326,9 +324,10 @@ def send_evicts(options):
-    ):
-        return True
+
+    # if options.cpu_type == "DerivO3CPU" or get_isa_from_str(args.isa) in (
+    #    ISA.X86,
+    #    ISA.ARM,
+    # ):
```
We need to know here if the target ISA is X86 or ARM.

BobbyRBruce commented:
@giactra I've pushed a new solution to this problem.

I've dropped `--isa` as an input parameter and now just have the user set the CPU Types by passing the class name. This class name is used to construct a CPU core via `ObjectList` and can be used to infer the target ISA of the system. It does this by using the logic that a system targeting ISA Y will have CPUs that are instances of that ISA CPU's Base class. E.g.,:  `is_x86_sim = isinstanceof(ObjectList.get(arg.cpu_type), X86CPU)`. So if user passes `X86TimingSimpleCPU` the scripts would infer an `X86` ISA where needed.

Let me know if this works for you.

giactra reviewed (changes requested):
> @giactra I've pushed a new solution to this problem.
> 
> I've dropped `--isa` as an input parameter and now just have the user set the CPU Types by passing the class name. This class name is used to construct a CPU core via `ObjectList` and can be used to infer the target ISA of the system. It does this by using the logic that a system targeting ISA Y will have CPUs that are instances of that ISA CPU's Base class. E.g.,: `is_x86_sim = isinstanceof(ObjectList.get(arg.cpu_type), X86CPU)`. So if user passes `X86TimingSimpleCPU` the scripts would infer an `X86` ISA where needed.
> 
> Let me know if this works for you.

Thank you so much Bobby, this clearly looks like a step in the right direction. I believe we are almost there, there's only one improvement I would like to suggest:

Rather than always relying on the cpu type passed with the command line ->`ObjectList.get_isa(args.cpu_type)`, I would *try* to extrapolate the ISA from the instantiated cpu object itself *whenever it is possible*.

For example if every ArchCPU  (ArmCPU,RiscvCPU, etc) defined a get_isa as a classmethod, I would prefer seeing something like:

```
if system.cpu.get_isa() in ['ISA.X86', ''ISA.ARM]:
    # Do x86 and ARM stuff
```

This IMHO has the following advantages:
1) You wouldn't be required to import ObjectList if you didn't want to
2) It is more readable
3) It works better when you instantiate CPUs with multiple ISAs

Of course I would still keep the current implementation whenever the CPU object is not yet available... But I would also consider re-implementing the get_isa in ObjectList by calling the cpu_type.get_isa directly (it would be a classmethod afterall).

Please let me know if this looks reasonable for you

BobbyRBruce commented:
> > @giactra I've pushed a new solution to this problem.
> > I've dropped `--isa` as an input parameter and now just have the user set the CPU Types by passing the class name. This class name is used to construct a CPU core via `ObjectList` and can be used to infer the target ISA of the system. It does this by using the logic that a system targeting ISA Y will have CPUs that are instances of that ISA CPU's Base class. E.g.,: `is_x86_sim = isinstanceof(ObjectList.get(arg.cpu_type), X86CPU)`. So if user passes `X86TimingSimpleCPU` the scripts would infer an `X86` ISA where needed.
> > Let me know if this works for you.
> 
> Thank you so much Bobby, this clearly looks like a step in the right direction. I believe we are almost there, there's only one improvement I would like to suggest:
> 
> Rather than always relying on the cpu type passed with the command line ->`ObjectList.get_isa(args.cpu_type)`, I would _try_ to extrapolate the ISA from the instantiated cpu object itself _whenever it is possible_.
> 
> For example if every ArchCPU (ArmCPU,RiscvCPU, etc) defined a get_isa as a classmethod, I would prefer seeing something like:
> 
> ```
> if system.cpu.get_isa() in ['ISA.X86', ''ISA.ARM]:
>     # Do x86 and ARM stuff
> ```
> 
> This IMHO has the following advantages:
> 
> 1. You wouldn't be required to import ObjectList if you didn't want to
> 2. It is more readable
> 3. It works better when you instantiate CPUs with multiple ISAs
> 
> Of course I would still keep the current implementation whenever the CPU object is not yet available... But I would also consider re-implementing the get_isa in ObjectList by calling the cpu_type.get_isa directly (it would be a classmethod afterall).
> 
> Please let me know if this looks reasonable for you

Would it be ok with you if I log this is an enhancement to add soon and we go with this solution for now? I really do want there to be better support for determining an CPU's ISA. In fact I'd quite look forward to working on that, but it seems a little out-of-scope for this PR which is just to remove a deprecated function. A change, like the one you propose, would allow us to refactor a lot of funny logic away, which is great, but i worry this PR is starting to do too much. Right now this PR removes `get_runtime_isa` with only a little additional code in the config scripts where knowing a Core's ISA is required. The necessity for this change was only to get this scripts working in the multi-ISA complication scenario, i think we've achieved that.

rogerchang23424 commented on configs/example/hmc_hello.py:
```diff
@@ -36,16 +36,23 @@
+    "--cpu-type",
+    type=str,
+    default="X86TimingSimpleCPU",
+    choices=CPUList.get_names(),
```
```
NameError: name 'CPUList' is not defined
```
Please fix

BobbyRBruce replied:
Thanks @rogerchang23424 ! Fixed!

BobbyRBruce commented:
> > > @giactra I've pushed a new solution to this problem.
> > > I've dropped `--isa` as an input parameter and now just have the user set the CPU Types by passing the class name. This class name is used to construct a CPU core via `ObjectList` and can be used to infer the target ISA of the system. It does this by using the logic that a system targeting ISA Y will have CPUs that are instances of that ISA CPU's Base class. E.g.,: `is_x86_sim = isinstanceof(ObjectList.get(arg.cpu_type), X86CPU)`. So if user passes `X86TimingSimpleCPU` the scripts would infer an `X86` ISA where needed.
> > > Let me know if this works for you.
> > 
> > 
> > Thank you so much Bobby, this clearly looks like a step in the right direction. I believe we are almost there, there's only one improvement I would like to suggest:
> > Rather than always relying on the cpu type passed with the command line ->`ObjectList.get_isa(args.cpu_type)`, I would _try_ to extrapolate the ISA from the instantiated cpu object itself _whenever it is possible_.
> > For example if every ArchCPU (ArmCPU,RiscvCPU, etc) defined a get_isa as a classmethod, I would prefer seeing something like:
> > ```
> > if system.cpu.get_isa() in ['ISA.X86', ''ISA.ARM]:
> >     # Do x86 and ARM stuff
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This IMHO has the following advantages:
> > 
> > 1. You wouldn't be required to import ObjectList if you didn't want to
> > 2. It is more readable
> > 3. It works better when you instantiate CPUs with multiple ISAs
> > 
> > Of course I would still keep the current implementation whenever the CPU object is not yet available... But I would also consider re-implementing the get_isa in ObjectList by calling the cpu_type.get_isa directly (it would be a classmethod afterall).
> > Please let me know if this looks reasonable for you
> 
> Would it be ok with you if I log this is an enhancement to add soon and we go with this solution for now? I really do want there to be better support for determining an CPU's ISA. In fact I'd quite look forward to working on that, but it seems a little out-of-scope for this PR which is just to remove a deprecated function. A change, like the one you propose, would allow us to refactor a lot of funny logic away, which is great, but i worry this PR is starting to do too much. Right now this PR removes `get_runtime_isa` with only a little additional code in the config scripts where knowing a Core's ISA is required. The necessity for this change was only to get this scripts working in the multi-ISA complication scenario, i think we've achieved that.

I've added the issue here: #622

giactra reviewed (approved):
Makes sense. Do you mind if I take care of the proposed improvement? Might save us some back and forth. Thanks for this fix Bobby!

powerjg commented:
@giactra I'll speak for Bobby and say you're welcome to add additional improvements 😄.

Do you want us to hold off merging until you push additional improvements?

giactra commented:
> 
> Do you want us to hold off merging until you push additional improvements?

No need, as Bobby mentioned in his comment, this PR already does its job in fixing a real bug; I will post my suggested enhancements separately targeting next release

abmerop commented:
Anyone else experiencing this breaking the send_evicts function in Ruby.py?  GPU SE mode is broken with this fix and I don't know what the problem is.  Wondering if it's happening for other Ruby protocols

mattsinc commented:
> Anyone else experiencing this breaking the send_evicts function in Ruby.py? GPU SE mode is broken with this fix and I don't know what the problem is. Wondering if it's happening for other Ruby protocols

I guess we would have seen this fail the daily GPU test then?

mattsinc commented:
> This seems fine from a GPU perspective, but just making sure you checked that at least the daily tests pass with it?

Sorry @abmerop , I should have followed up on this question previously, as it appears the answer was no.
